MuJoCo 2.1 compatible gym_compete

PavelCz opened the pull request:
- Update to muojco-py 2.1 which is the latest version and supports MuJoCo 2.1
- Remove copy of `mujoco_env` from `openai/gym` in favor of using the `openai/gym` version directly
- Remove `policy.py`. Not sure if this is desired, the main reason is that the current `policy.py `is only compatible with `tf`<2 and would need to be reworked. Older versions of `tf` might also limit the version of `gym` which could be used

PavelCz committed:
Re-introduce main.py for running demo tasks

PavelCz committed:
Update registered name of env to new correctly formatted name

in demo code in main.py, only for YouShallNotPass

PavelCz committed:
Re-add get_variables() in GymCompetePolicy. Is this necessary? Used in main.py

PavelCz committed:
Start converting main.py to mujoco-py 2.1 and tf2. Unfinished!

PavelCz committed:
Start fixing demo_tasks.sh

PavelCz committed:
Remove main.py/demo script. Too many changes in other code.

This makes get_variables() in policy.py superfluous, which is also removed

PavelCz committed:
Port code to newer version of mujoco-py (/ MuJoCo)

It looks like many things that used to be in model are now in sim. Use strings instead of byte sequences to for keys

PavelCz committed:
Update to tensorflow 2, use backward compatibility packages

PavelCz committed:
Update requirements

PavelCz committed:
Fix simulation reset

PavelCz committed:
The numarrows attribute for visual quality was removed in MuJoCo 2.0

This setting was removed from the XML Schema in 2.0, because it was not actually used, so I removed it from the xml asset

PavelCz committed:
Change xmat to body_xmat attribute

I couldn't find where, but apparently at some point this attribute was changed

PavelCz committed:
Update sumo to mujoco-py 2.1

PavelCz committed:
Update humanoid kicker agent and kick_and_defend to mujoco-py 2.1

PavelCz committed:
Update readme and requirements

PavelCz committed:
Update setup.py

PavelCz committed:
Update stable-baselines requirement with specific commit

PavelCz committed:
Update README

PavelCz committed:
Remove policy.py and with that the dependency on stable-baselines

PavelCz committed:
Fix viewer camera moving upwards each time env is reset

PavelCz committed:
Update setup

PavelCz committed:
Remove unused import that I accidentally added previously

PavelCz committed:
Use mujoco env from gym instead of old copy

PavelCz committed:
Remove unused imports

PavelCz committed:
Remove clipping actions before rew calculation for humanoid agents

Other agents didn't have this clipping in the first place, so needed no change

PavelCz committed:
Debug huge actions

PavelCz committed:
Don't silently ignore NaN, don't catch and immediately throw exception

PavelCz committed:
Remove clipping from multi_agent_env

PavelCz committed:
Merge branch 'no-clip'

PavelCz committed:
Update readme

PavelCz committed:
Update setup.py and increase version number

PavelCz committed:
Changes for upstream pull request

PavelCz commented:
Looks like I can't add you as a reviewer @AdamGleave.

PavelCz committed:
Remove commented camera code in multi_agent_scene

PavelCz committed:
Remove commented camera code in multi_agent_scene

PavelCz committed:
Remove unused import

PavelCz committed:
Remove unused import

PavelCz committed:
Add changes to readme

PavelCz commented:
Reminder to not merge this into `master`. AFAICT there is no way for me to set up the PR such that it gets merged into a new brach that doesn't exist yet, but I assume as admin you have that ability.

AdamGleave commented on README.md:
```diff
@@ -1,16 +1,28 @@
+  - `env.model` -> `env.sim`
+  - Some attributes were renamed.
+- Update for compatiblity with newer version of `numpy`.
+- Some agents clipped actions before calculating rewards, some didn't. Now there is consistently no clipping.
```
```suggestion
- Some agents clipped actions before calculating rewards, some didn't. Now there is consistently no clipping in this codebase, although it is still clipped to actuator bounds inside `MujocoEnv`.
```

AdamGleave commented on README.md:
```diff
@@ -1,16 +1,28 @@
+- Update for compatiblity with newer version of `numpy`.
+- Some agents clipped actions before calculating rewards, some didn't. Now there is consistently no clipping.
+- The `numarrows` attribute in xml was removed due to it having no function.
+- Remove copied code of `openai/gym`'s `mujoco_env` in favor of importing from gym.
```
```suggestion
- Remove copied code of `openai/gym`'s `mujoco_env` in favor of importing from `gym`.
```

AdamGleave commented on README.md:
```diff
@@ -1,16 +1,28 @@
+- The `numarrows` attribute in xml was removed due to it having no function.
+- Remove copied code of `openai/gym`'s `mujoco_env` in favor of importing from gym.
+- Remove `policy_py` which relies on `tensorflow<2`
+- Remove the `demo_tasks.sh` script which wasn't runnable due to prior changes in the HumanCompatibleAI fork.
```
Can you cherry-pick this commit, make a separate PR, and then we'll merge that to master (and then merge this PR to a separate branch)? We shouldn't have unrannable code in master either...

PavelCz replied:
Done

AdamGleave commented on README.md:
````diff
@@ -19,15 +31,10 @@ Next install `gym-compete` package as:
+Or use pip to install / add to requirements as
 ```bash
-bash demo_tasks.sh all
+gym_compete @ git+https://github.com/HumanCompatibleAI/multiagent-competition.git@v0.1.0
````
Will this install the right branch? I think we might need something like `@mujoco21` (if new branch is called `mujoco21`)

PavelCz replied:
Yes, you're right that's what it should be called.

AdamGleave commented on README.md:
````diff
@@ -19,15 +31,10 @@ Next install `gym-compete` package as:
-bash demo_tasks.sh <task>
-```
-where `<task>` is one of: `run-to-goal-humans`, `run-to-goal-ants`, `you-shall-not-pass`, `sumo-ants`, `sumo-humans` and `kick-and-defend`
+Check install is successful by coming out of the directory and trying `import gym_compete` in python console. Some users might require a `sudo pip install`.
````
```suggestion
Check install is successful by coming out of the directory and trying `import gym_compete` in python console.
```
They should not if they're using virtualenvs correctly...

AdamGleave commented on README.md:
```diff
@@ -1,16 +1,28 @@
+- Some agents clipped actions before calculating rewards, some didn't. Now there is consistently no clipping.
+- The `numarrows` attribute in xml was removed due to it having no function.
+- Remove copied code of `openai/gym`'s `mujoco_env` in favor of importing from gym.
+- Remove `policy_py` which relies on `tensorflow<2`
```
```suggestion
- Remove `policy.py` which relies on `tensorflow<2`
```

AdamGleave commented on gym_compete/new_envs/agents/humanoid.py:
```diff
@@ -4,7 +4,8 @@
 
 def mass_center(mass, xpos):
-    return (np.sum(mass * xpos, 0) / np.sum(mass))[0]
+    # I think the reshape is necessary because np got stricter about broadcasting dims of size 1 vs dims of size 0, if not this calculation might be wrong
```
This comment doesn't inspire confidence :) Following line looks correct to me, if you're not sure can you test it a bit?

PavelCz replied:
I investigated this, see comment below

AdamGleave reviewed (changes requested):
Thanks for the PR, some minor changes suggested.

I've updated the base branch (had to push new branch first to make it an option, but I don't think you have write access to this repo so you wouldn't have been able to do that).

PavelCz committed:
Clarify README.md

Co-authored-by: Adam Gleave <adam@gleave.me>

PavelCz committed:
Update README.md

Co-authored-by: Adam Gleave <adam@gleave.me>

PavelCz committed:
Update README.md

Co-authored-by: Adam Gleave <adam@gleave.me>

PavelCz committed:
Update README.md

Co-authored-by: Adam Gleave <adam@gleave.me>

PavelCz committed:
Fix faulty render call

The argument 'close' for gym's mujoco_env was removed in commit https://github.com/openai/gym/commit/4c460ba6c8959dd8e0a03b13a1ca817da6d4074f .
We weren't using this option, but it made it so the render function was called with wrong arguments.

PavelCz committed:
Remove unrunnable demo script

In the following commit main.py was removed, consequently demo_task.sh can't be run anymore: 60ee87dbb24d814986a4c045891b7a20f845fe81

PavelCz committed:
Actually remove unrunnable demo script

Actually do what the previous commit message claimed, that is:
In the following commit main.py was removed, consequently demo_task.sh can't be run anymore: 60ee87dbb24d814986a4c045891b7a20f845fe81

PavelCz committed:
Merge branch 'upstream-master' into changes-for-upstream

PavelCz committed:
Change branch to install from

PavelCz committed:
Use idiomatic way of adding axis, correct comment in mass center calc

PavelCz commented:
For posterity:
The changes in `mass_center()` are necessary because in some version prior to mujoco 2.1 `body_mass` was changed from 1d matrix (shape (13,1)) to vector (shape (13,)). Multiplying (`*`) this with a matrix (shape (13,3)) will fail on the vector. The reason is that numpy considers vectors as [1,n] arrays as opposed to [n,1] arrays during broadcasting.

PavelCz committed:
Merge branch 'master' into changes-for-upstream

PavelCz commented:
@AdamGleave I addressed all comments.
Stephen also found a bug which I fixed in [this commit](https://github.com/HumanCompatibleAI/multiagent-competition/pull/6/commits/3620b978d6ee88699b46a3bcf2d9d1282fca8594) (now also part of this PR).

PavelCz committed:
Remove unrunnable demo script (#7)

* Remove unrunnable demo script

In the following commit main.py was removed, consequently demo_task.sh can't be run anymore: 60ee87dbb24d814986a4c045891b7a20f845fe81

* Actually remove unrunnable demo script

Actually do what the previous commit message claimed, that is:
In the following commit main.py was removed, consequently demo_task.sh can't be run anymore: 60ee87dbb24d814986a4c045891b7a20f845fe81

PavelCz committed:
Merge remote-tracking branch 'upstream/master' into upstream-master

PavelCz committed:
Merge branch 'upstream-master' into changes-for-upstream

AdamGleave reviewed (approved):
LGTM
